# Mall Customer Segmentation & Clustering Analysis

## 1. Introduction 

### 1.1 Project Description

Customer segmentation is a fundamental strategy in modern retail and marketing. This project performs an in-depth analysis of customer data from a shopping mall to identify distinct customer groups based on their purchasing behavior and demographic characteristics.

By leveraging unsupervised machine learning techniques, we aim to uncover hidden patterns in customer behavior that would otherwise go unnoticed. These patterns serve as the foundation for targeted marketing strategies, personalized customer experiences, and optimized inventory management.

This analysis employs **3 different clustering algorithms**, comparing their strengths and weaknesses to identify the most effective segmentation approach for mall customer data.


### 1.2 Project Objectives

**Primary Objectives:**

1. **Identify Optimal Customer Clusters**
   - Determine the ideal number of customer segments using Elbow Method and Silhouette Analysis
   - Create distinct, actionable customer groups

2. **Implement Multiple Clustering Algorithms**
   - Compare 3 different unsupervised learning approaches
   - Evaluate which algorithm best fits mall customer data
   - Understand strengths/weaknesses of each method

3. **Profile Customer Segments**
   - Analyze demographic and behavioral characteristics of each cluster
   - Create detailed customer personas for each segment
   - Understand purchasing power and spending patterns

4. **Generate Actionable Business Insights**
   - Develop marketing strategies tailored to each segment
   - Identify high-value vs. at-risk customer groups
   - Provide recommendations for business growth



### 1.3 Dataset Overview

**Dataset Source:** Mall Customer Segmentation Dataset

**Dataset Statistics:**
| Aspect | Details |
|--------|---------|
| **Total Records** | 1,000 customers |
| **Total Features** | 5 columns |
| **Data Type** | Structured, Numerical + Categorical |
| **Missing Values** | Minimal (< 1%) |
| **Format** | CSV |

**Features Included:**

1. **CustomerID** (Identifier)
   - Unique customer identifier
   - Range: 1 - 1000
   - Use: Row identification only (dropped in analysis)

2. **Gender** (Categorical)
   - Customer gender
   - Values: Male, Female
   - Use: Segment profiling

3. **Age** (Numerical)
   - Customer age in years
   - Range: 18 - 70 years
   - Distribution: Economically active population
   - Use: Demographic segmentation

4. **Annual Income** (Numerical)
   - Customer annual income in thousands ($k)
   - Range: 15k - 137k USD
   - Distribution: Realistic career progression patterns
   - Use: Primary clustering feature

5. **Spending Score** (Numerical)
   - Behavioral metric (1-100 scale)
   - Assigned by mall based on customer behavior, purchase history, frequency
   - Range: 1 - 100
   - Distribution: Behavioral variation across segments
   - Use: Primary clustering feature

**Key Insights from Dataset:**
- Clean, pre-processed data (minimal missing values)
- Balanced distribution across demographics
- Real-world applicability for retail businesses
- Suitable for unsupervised learning tasks

### 1.4 Clustering Algorithms to Be Implemented

This project will implement and compare **3 different unsupervised clustering algorithms:**

#### **1. K-Means Clustering**
- **Type:** Partition-based, centroid-based
- **How it works:** Divides data into K clusters by minimizing within-cluster variance
- **Advantages:** Fast, scalable, simple to understand
- **Best for:** Well-separated spherical clusters


#### **2. Hierarchical Clustering (Agglomerative)**
- **Type:** Hierarchical, bottom-up approach
- **How it works:** Builds dendrograms by merging closest clusters
- **Advantages:** Interpretable, no need to specify K beforehand
- **Best for:** Understanding cluster relationships

#### **3. DBSCAN (Density-Based Spatial Clustering)**
- **Type:** Density-based
- **How it works:** Groups points based on density and proximity
- **Advantages:** Can find arbitrary-shaped clusters, identifies outliers
- **Best for:** Non-spherical clusters, noisy data


## 2. Data Loading & Inspection

### 2.1 Introduction

Before applying any machine learning algorithms, it is essential to understand the structure, format, and quality of the dataset. This section focuses on:

- Loading the dataset into memory
- Inspecting the basic structure (rows, columns, data types)
- Understanding the data characteristics
- Identifying any data quality issues (missing values, inconsistencies)
- Preparing for subsequent analysis steps


### 2.2 Import Required Libraries

We begin by importing all necessary libraries for data manipulation, analysis, and visualization.

In [1]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns

import plotly.graph_objects as go
import plotly.express as px
from plotly.subplots import make_subplots

import plotly.io as pio
pio.renderers.default = "notebook_connected"

from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans, AgglomerativeClustering, DBSCAN
from sklearn.metrics import silhouette_score, silhouette_samples, calinski_harabasz_score, davies_bouldin_score

from scipy.cluster.hierarchy import dendrogram, linkage
from scipy.spatial.distance import cdist

import skfuzzy as fuzz
import warnings
warnings.filterwarnings('ignore')

### 2.3 Load Dataset

Load the customer data from the CSV file.

In [2]:
df = pd.read_csv('store_customers.csv')

df.head()

,CustomerID,Gender,Age,Annual Income (k$),Spending Score (1-100)
0,1000,M,39.0,59.9,58.0
1,1001,M,34.0,48.4,37.0
2,1002,F,40.0,70.5,26.0
3,1003,F,47.0,81.1,30.0
4,1004,F,33.0,42.1,58.0


In [3]:
df.shape

(1000, 5)

In [ ]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   CustomerID              1000 non-null   int64  
 1   Gender                  997 non-null    str    
 2   Age                     994 non-null    float64
 3   Annual Income (k$)      996 non-null    float64
 4   Spending Score (1-100)  994 non-null    float64
dtypes: float64(3), int64(1), str(1)
memory usage: 39.2 KB


In [5]:
df.describe()

,CustomerID,Age,Annual Income (k$),Spending Score (1-100)
count,1000.000000,994.000000,996.000000,994.000000
mean,1499.500000,38.935614,57.149096,42.645875
std,288.819436,13.399880,28.628506,20.101589
min,1000.000000,18.000000,15.000000,1.000000
25%,1249.750000,30.000000,34.975000,31.000000
50%,1499.500000,36.000000,49.000000,47.000000
75%,1749.250000,44.000000,79.400000,57.000000
max,1999.000000,80.000000,144.100000,92.000000


### 2.7 Missing Values Summary

Identify and quantify any missing values in the dataset.

In [6]:
missing_values = df.isnull().sum()
print("Missing Values Count:")
print(missing_values)
print("\nMissing Values Percentage:")
print((df.isnull().sum() / len(df)) * 100)

Missing Values Count:
CustomerID                0
Gender                    3
Age                       6
Annual Income (k$)        4
Spending Score (1-100)    6
dtype: int64

Missing Values Percentage:
CustomerID                0.0
Gender                    0.3
Age                       0.6
Annual Income (k$)        0.4
Spending Score (1-100)    0.6
dtype: float64


In [7]:
print("DATA QUALITY REPORT")
print("-" * 60)
print(f"Completeness: {(1 - df.isnull().sum().sum() / (df.shape[0] * df.shape[1])) * 100:.2f}%")
print(f"Duplicate Records: {df.duplicated().sum()}")
print(f"Outliers (to be assessed in EDA): Pending")
print("-" * 60)

DATA QUALITY REPORT
------------------------------------------------------------
Completeness: 99.62%
Duplicate Records: 0
Outliers (to be assessed in EDA): Pending
------------------------------------------------------------


## 3. Data Cleaning & Preprocessing

### 3.1 Introduction

Data cleaning is the first critical step in any data science project. Raw data often contains inconsistencies, missing values, and formatting issues that must be addressed before analysis.

In this section, we will:
- Remove rows with missing values
- Rename columns for clarity
- Remove unnecessary columns
- Encode categorical variables
- Prepare data for analysis



In [8]:
df.isna().sum()

CustomerID                0
Gender                    3
Age                       6
Annual Income (k$)        4
Spending Score (1-100)    6
dtype: int64

In [9]:
df.dropna(inplace=True)
df.shape

(982, 5)

In [10]:
df = df.rename(columns={
    "Annual Income (k$)": "AnnualIncome",
    "Spending Score (1-100)": "SpendingScore"
})

In [11]:
df.columns

Index(['CustomerID', 'Gender', 'Age', 'AnnualIncome', 'SpendingScore'], dtype='str')

In [12]:
df = df.drop(columns=["CustomerID"])
df.shape

(982, 4)

In [13]:
df.head()

,Gender,Age,AnnualIncome,SpendingScore
0,M,39.0,59.9,58.0
1,M,34.0,48.4,37.0
2,F,40.0,70.5,26.0
3,F,47.0,81.1,30.0
4,F,33.0,42.1,58.0


### 3.6 Encode Categorical Variables

The Gender column contains text values (M, F). For clustering, we need to convert these to numbers.
**Mapping:**
- M (Male) → 1
- F (Female) → 0

In [14]:
df["Gender"] = df["Gender"].map({'M':1, 'F':0})

In [15]:
df["Gender"].value_counts()

Gender
0    511
1    471
Name: count, dtype: int64

In [16]:
print("=" * 60)
print("CLEANED DATA SUMMARY")
print("=" * 60)
print(f"Total Records: {df.shape[0]}")
print(f"Total Features: {df.shape[1]}")
print(f"Missing Values: {df.isnull().sum().sum()}")
print("\nColumn Names and Types:")
print(df.dtypes)
print("\nBasic Statistics:")
print(df.describe())
print("=" * 60)

CLEANED DATA SUMMARY
Total Records: 982
Total Features: 4
Missing Values: 0

Column Names and Types:
Gender             int64
Age              float64
AnnualIncome     float64
SpendingScore    float64
dtype: object

Basic Statistics:
           Gender         Age  AnnualIncome  SpendingScore
count  982.000000  982.000000    982.000000     982.000000
mean     0.479633   38.958248     57.326782      42.603870
std      0.499840   13.417748     28.658961      20.121469
min      0.000000   18.000000     15.000000       1.000000
25%      0.000000   30.000000     35.125000      31.000000
50%      0.000000   36.000000     49.150000      47.000000
75%      1.000000   44.000000     79.475000      57.000000
max      1.000000   80.000000    144.100000      92.000000


## 4. Exploratory Data Analysis (EDA)

## 4.1 Introduction

Exploratory Data Analysis (EDA) is the process of visually and
statistically examining a dataset to discover patterns, understand
distributions, and uncover relationships between variables —
before applying any machine learning algorithm.

In this section, we will:
- Analyze distribution of Age, Annual Income, and Spending Score
- Examine gender breakdown across customers
- Identify correlations between features
- Detect outliers that may affect clustering

In [17]:
import numpy as np
import plotly.graph_objects as go
from scipy.stats import gaussian_kde

def plot_distribution(df, col, bins=30):
    
    data = df[col].dropna()

    hist = go.Histogram(
        x=data,
        nbinsx=bins,
        histnorm='probability density',
        marker=dict(
            color='#3498DB',
            line=dict(color='black', width=1.5)
        ),
        name='Histogram'
    )

    fig = go.Figure([hist])

    if data.dtype != 'object' and data.nunique() > 2:
        kde = gaussian_kde(data)

        x_range = np.linspace(data.min(), data.max(), 200)
        y_kde = kde(x_range)

        fig.add_trace(
            go.Scatter(
                x=x_range,
                y=y_kde,
                mode='lines',
                line=dict(color='red', width=2),
                name='KDE Curve'
            )
        )

    fig.update_layout(
        title=f'Distribution of {col}',
        xaxis_title=col,
        yaxis_title='Density',
        width=650,
        height=400,
        template='plotly_white'
    )

    fig.show()

In [18]:
plot_distribution(df, 'Age')

In [19]:
plot_distribution(df, 'AnnualIncome')

In [20]:
plot_distribution(df, 'SpendingScore')

In [21]:
import plotly.graph_objects as go

def plot_scatter(df, x_col, y_col, color='#3498DB', title=None):

    fig = go.Figure()

    fig.add_trace(go.Scatter(
        x=df[x_col],
        y=df[y_col],
        mode='markers',
        marker=dict(
            color=color,
            size=7,
            line=dict(color='black', width=0.5)
        ),
        name='Customers',
        hovertemplate=f'<b>{x_col}:</b> %{{x}}<br><b>{y_col}:</b> %{{y}}<extra></extra>'
    ))

    fig.update_layout(
        title=title if title else f'{x_col} vs {y_col}',
        xaxis_title=x_col,
        yaxis_title=y_col,
        width=640,
        height=400,
        template='plotly_white'
    )

    fig.show()

In [22]:
plot_scatter(df, 'AnnualIncome', 'SpendingScore', color='#3498DB')

In [23]:
plot_scatter(df, 'Age', 'SpendingScore', color='#E74C3C')

In [24]:
X = df[["AnnualIncome","SpendingScore"]].values
X

array([[ 59.9,  58. ],
       [ 48.4,  37. ],
       [ 70.5,  26. ],
       ...,
       [ 67.7,  26. ],
       [ 45.7,  59. ],
       [107.8,   1. ]], shape=(982, 2))

In [25]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)
X_scaled

array([[ 0.0898333 ,  0.76554926],
       [-0.31164181, -0.27864393],
       [ 0.45988863, -0.82560227],
       ...,
       [ 0.36213816, -0.82560227],
       [-0.40590119,  0.81527275],
       [ 1.76206444, -2.06868941]], shape=(982, 2))

## 5. Model Selection

### 5.1 K-Means Clustering

K-Means is a partition-based clustering algorithm that divides
data into K clusters by minimizing within-cluster variance. It is
the most widely used clustering algorithm in industry.

**How it works:**
- Randomly initialize K centroids
- Assign each point to the nearest centroid
- Recalculate centroids based on assigned points
- Repeat until convergence

### 5.2 Finding the Optimal Number of Clusters

Before applying K-Means, we need to determine the ideal value of K.
We use two methods for this:

### 5.3 Elbow Method

The Elbow Method plots inertia (within-cluster sum of squares)
against different values of K. The point where the curve bends
like an "elbow" indicates the optimal number of clusters.

**How it works:**
- Run K-Means for K = 1 to 10
- Calculate inertia for each K
- Plot inertia vs K
- Identify the elbow point where inertia stops decreasing significantly

In [26]:
from sklearn.cluster import KMeans

inertia = []

K = range(1,11)
for k in K:
    kmeans = KMeans(n_clusters=k, random_state=42,n_init=10)
    kmeans.fit(X_scaled)
    inertia.append(kmeans.inertia_)

inertia

[1963.9999999999995,
 627.4437722304631,
 380.70474879676004,
 277.4085012407078,
 227.12730374808604,
 193.0142741792354,
 160.27523541821773,
 141.9284308202654,
 127.74877627093808,
 115.97961447218836]

In [27]:
fig = go.Figure()

fig.add_trace(go.Scatter(
    x=list(K),
    y=inertia,
    mode='lines+markers',
    marker=dict(
        color='#E74C3C',
        size=8,
        line=dict(color='black', width=1)
    ),
    line=dict(color='#3498DB', width=2),
    hovertemplate='<b>K:</b> %{x}<br><b>Inertia:</b> %{y:.2f}<extra></extra>'
))

fig.update_layout(
    title='Elbow Method - Optimal K Selection',
    xaxis_title='Number of Clusters (K)',
    yaxis_title='Inertia',
    width=645,
    height=400,
    template='plotly_white'
)
fig.show()

### 5.4 Silhouette Analysis

Silhouette Analysis measures how similar a point is to its own
cluster compared to other clusters. Score ranges from -1 to 1 —
higher is better.

**How it works:**
- Run K-Means for K = 2 to 10
- Calculate silhouette score for each K
- Higher score = better defined clusters
- Select K with the highest silhouette score

In [28]:
from sklearn.metrics import silhouette_score

sil_scores = []

K = range(2,11)
for k in K:
    kmeans = KMeans(n_clusters=k,random_state=42, n_init=10)
    labels = kmeans.fit_predict(X_scaled)
    score = silhouette_score(X_scaled, labels)
    sil_scores.append(score)

sil_scores

[0.5882647596002428,
 0.474956222630729,
 0.3957499299556574,
 0.36362726099229903,
 0.3634749904887551,
 0.3732245755293607,
 0.36009775587028675,
 0.35410879741907947,
 0.3537239290234216]

In [29]:
import plotly.graph_objects as go

fig = go.Figure()

fig.add_trace(
    go.Scatter(
        x=list(K),
        y=sil_scores,
        mode='lines+markers',
        line=dict(color='deepskyblue', width=3),
        marker=dict(size=10, color='red'),
        name='Silhouette Score'
    )
)

fig.update_layout(
    title='Silhouette Analysis',
    xaxis_title='Number of Clusters (k)',
    yaxis_title='Silhouette Score',
    template='plotly_white',
    hovermode='x unified',
    width=645,
    height=400,
)

fig.show()

### 5.5 Decision — Final K Selection

Based on both the Elbow Method and Silhouette Analysis, we select
the optimal number of clusters.

**Elbow Method:** Curve bends at K = 5, indicating diminishing
returns beyond this point.

**Silhouette Analysis:** K = 5 yields the highest silhouette score,
confirming well-defined and separated clusters.

**Final Decision: K = 5**

K = 5 will be used for all subsequent clustering algorithms.

## 6. K-Means Clustering

### 6.1 Model Training

K-Means model is trained with K = 5 on the scaled data.



In [30]:
kmeans = KMeans(n_clusters=5, random_state=42, n_init=10, init='k-means++')
labels_kmeans = kmeans.fit_predict(X_scaled)

In [31]:
df["KMeans_Cluster"] = labels_kmeans
df.head()

,Gender,Age,AnnualIncome,SpendingScore,KMeans_Cluster
0,1,39.0,59.9,58.0,1
1,1,34.0,48.4,37.0,1
2,0,40.0,70.5,26.0,2
3,0,47.0,81.1,30.0,2
4,0,33.0,42.1,58.0,3


### 6.2 Cluster Visualization

We visualize the K-Means clusters by plotting AnnualIncome vs
SpendingScore, with each cluster colored differently.

In [32]:
import plotly.graph_objects as go
import numpy as np

fig = go.Figure()

# Clusters
for cluster in np.unique(labels_kmeans):
    fig.add_trace(
        go.Scatter(
            x=X[labels_kmeans == cluster, 0],
            y=X[labels_kmeans == cluster, 1],
            mode='markers',
            name=f'Cluster {cluster}',
            marker=dict(size=8),
            hovertemplate=f'<b>Cluster {cluster}</b><br><b>Annual Income:</b> %{{x}}k$<br><b>Spending Score:</b> %{{y}}<extra></extra>'
        )
    )

# Centroids
centers = scaler.inverse_transform(kmeans.cluster_centers_)

fig.add_trace(
    go.Scatter(
        x=centers[:, 0],
        y=centers[:, 1],
        mode='markers',
        name='Centroids',
        marker=dict(
            symbol='x',
            size=18,
            color='black',
            line=dict(width=2)
        ),
        hovertemplate='<b>Centroid</b><br><b>Annual Income:</b> %{x}k$<br><b>Spending Score:</b> %{y}<extra></extra>'
    )
)

fig.update_layout(
    title='K-Means Clustering (K=5)',
    xaxis_title='Annual Income',
    yaxis_title='Spending Score',
    template='plotly_white',
    hovermode='closest',
    width=700,
    height=400,
)

fig.show()

### 6.3 Cluster Analysis

We analyze each cluster to understand the characteristics and
behavior of customers in each segment.

In [33]:
df.groupby("KMeans_Cluster")[["AnnualIncome","SpendingScore","Age"]].mean()

,AnnualIncome,SpendingScore,Age
KMeans_Cluster,,,
0,29.803226,66.774194,28.016129
1,61.457812,42.244792,38.682292
2,84.554839,26.825806,46.619355
3,37.920433,51.368421,31.133127
4,107.915079,4.412698,66.166667


### 6.4 Customer Segmentation

Each cluster is ranked based on business value using
AnnualIncome and SpendingScore. Higher income and higher
spending = more valuable customer.

**BusinessScore Formula:**
- 50% AnnualIncome + 50% SpendingScore
- Age excluded (age does not determine customer value)

In [34]:
cluster_stats = df.groupby("KMeans_Cluster")[["AnnualIncome", "SpendingScore", "Age"]].mean().reset_index()

norm = cluster_stats.copy()
for col in ["AnnualIncome", "SpendingScore", "Age"]:
    norm[col] = (norm[col] - norm[col].min()) / (norm[col].max() - norm[col].min())

norm["BusinessScore"] = (
    0.50 * norm["AnnualIncome"] +
    0.50 * norm["SpendingScore"]
)

norm = norm.sort_values("BusinessScore", ascending=False).reset_index(drop=True)

labels = [
    ("⭐⭐⭐⭐⭐", "Best"),
    ("⭐⭐⭐⭐", "Potential"),
    ("⭐⭐⭐", "Sustainable"),
    ("⭐⭐", "High Risk"),
    ("⭐", "Low Value")
]

norm["Business Value"] = [l[0] for l in labels]
norm["Behavior"] = [l[1] for l in labels]

final_table = norm[["KMeans_Cluster", "Business Value", "Behavior"]] \
    .rename(columns={"KMeans_Cluster": "Cluster"})

final_table

,Cluster,Business Value,Behavior
0,2,⭐⭐⭐⭐⭐,Best
1,1,⭐⭐⭐⭐,Potential
2,0,⭐⭐⭐,Sustainable
3,4,⭐⭐,High Risk
4,3,⭐,Low Value


### 6.5 Cluster Characteristics Heatmap

Heatmap shows mean values of each feature per cluster.
Color intensity indicates the relative importance of each
feature — darker color means higher value.

In [35]:
cluster_means = df.groupby("KMeans_Cluster")[["Age", "AnnualIncome", "SpendingScore"]].mean().round(2)

fig = px.imshow(
    cluster_means,
    text_auto=True,
    color_continuous_scale='RdBu_r',
    title='Cluster Characteristics Heatmap',
    labels=dict(x='Features', y='Cluster', color='Mean Value')
)

fig.update_layout(
    xaxis_title='Features',
    yaxis_title='Cluster',
    width=500,
    height=500,
    template='plotly_white'
)

fig.show()

### 6.6 Box Plot — Features by Cluster

Box plots show the distribution of Age, AnnualIncome, and
SpendingScore across each cluster. This helps identify
spread and outliers within each segment.

In [36]:
features = ["Age", "AnnualIncome", "SpendingScore"]
colors = ['#3498DB', '#E74C3C', '#2ECC71', '#F39C12', '#9B59B6']

for feature in features:
    fig = go.Figure()

    for cluster in sorted(df["KMeans_Cluster"].unique()):
        fig.add_trace(go.Box(
            y=df[df["KMeans_Cluster"] == cluster][feature],
            name=f'Cluster {cluster}',
            marker=dict(
                color=colors[cluster],
                line=dict(color='black', width=1)
            ),
            boxpoints='all',
            jitter=0.3,
            pointpos=-1.8
        ))

    fig.update_layout(
        title=f'{feature} Distribution by Cluster',
        xaxis_title='Cluster',
        yaxis_title=feature,
        width=650,
        height=400,
        template='plotly_white'
    )

    fig.show()

## 7. Hierarchical Clustering (Agglomerative)

### 7.1 Introduction

Hierarchical Clustering builds a tree of clusters (dendrogram)
by progressively merging the closest data points. Unlike K-Means,
it does not require specifying K beforehand.

**How it works:**
- Start with each point as its own cluster
- Merge the two closest clusters
- Repeat until all points are in one cluster
- Cut the dendrogram at desired level to get K clusters

In [37]:
from sklearn.cluster import AgglomerativeClustering
from scipy.cluster.hierarchy import dendrogram, linkage

In [38]:
linked = linkage(X_scaled, method='ward')

In [39]:
from scipy.cluster.hierarchy import dendrogram, linkage

linked = linkage(X_scaled, method='ward')

dend = dendrogram(linked, truncate_mode='lastp', p=12, no_plot=True)

color_map = {
    'C0': '#3498DB',
    'C1': '#E74C3C',
    'C2': '#2ECC71',
    'C3': '#F39C12',
    'C4': '#9B59B6',
    'C5': '#1ABC9C',
    'b': '#3498DB',
    'g': '#2ECC71',
    'r': '#E74C3C',
    'k': '#2C3E50'
}

fig = go.Figure()

for icoord, dcoord, color in zip(dend['icoord'], dend['dcoord'], dend['color_list']):
    plotly_color = color_map.get(color, '#3498DB')
    fig.add_trace(go.Scatter(
        x=icoord,
        y=dcoord,
        mode='lines',
        line=dict(color=plotly_color, width=2),
        hovertemplate='<b>Distance:</b> %{y:.2f}<extra></extra>',
        showlegend=False
    ))

tick_text = [f"({dend['ivl'][i]})" for i in range(len(dend['ivl']))]

fig.update_layout(
    title='Hierarchical Clustering Dendrogram',
    xaxis=dict(
        tickvals=[5 + 10*i for i in range(len(tick_text))],
        ticktext=tick_text,
        title='Cluster Size'
    ),
    yaxis_title='Distance',
    template='plotly_white',
    width=700,
    height=400
)

fig.show()

In [40]:
hier = AgglomerativeClustering(
    n_clusters=5,
    linkage='ward'
)

labels_hier = hier.fit_predict(X_scaled)

In [41]:
df["Hierarchical_Cluster"] = labels_hier
df.head()

,Gender,Age,AnnualIncome,SpendingScore,KMeans_Cluster,Hierarchical_Cluster
0,1,39.0,59.9,58.0,1,1
1,1,34.0,48.4,37.0,1,1
2,0,40.0,70.5,26.0,2,3
3,0,47.0,81.1,30.0,2,4
4,0,33.0,42.1,58.0,3,1


In [42]:
fig = go.Figure()

for cluster in np.unique(labels_hier):
    fig.add_trace(
        go.Scatter(
            x=X[labels_hier == cluster, 0],
            y=X[labels_hier == cluster, 1],
            mode='markers',
            name=f'Cluster {cluster}',
            marker=dict(size=8),
            hovertemplate=f'<b>Cluster {cluster}</b><br><b>Annual Income:</b> %{{x}}k$<br><b>Spending Score:</b> %{{y}}<extra></extra>'
        )
    )

fig.update_layout(
    title='Hierarchical Clustering (K=5)',
    xaxis_title='Annual Income (k$)',
    yaxis_title='Spending Score (1-100)',
    template='plotly_white',
    hovermode='closest',
    width=645,
    height=400
)

fig.show()

## 8. DBSCAN Clustering


DBSCAN (Density-Based Spatial Clustering of Applications with Noise)
groups points based on density. Unlike K-Means, it does not require
specifying K and can detect outliers as noise points.

**Key Parameters:**
- **eps** — Maximum distance between two points to be considered neighbors
- **min_samples** — Minimum points required to form a dense region

**Advantages:**
- Finds arbitrary-shaped clusters
- Automatically detects outliers (noise points)
- No need to specify number of clusters

In [43]:
for eps in [0.1, 0.15, 0.2, 0.25, 0.3]:
    db = DBSCAN(eps=eps, min_samples=5)
    labels = db.fit_predict(X_scaled)
    n_clusters = len(set(labels)) - (1 if -1 in labels else 0)
    n_noise = list(labels).count(-1)
    print(f"eps={eps} | Clusters: {n_clusters} | Noise: {n_noise}")

eps=0.1 | Clusters: 30 | Noise: 212
eps=0.15 | Clusters: 2 | Noise: 65
eps=0.2 | Clusters: 2 | Noise: 26
eps=0.25 | Clusters: 1 | Noise: 14
eps=0.3 | Clusters: 1 | Noise: 4


In [44]:
dbscan = DBSCAN(eps=0.2, min_samples=5)
labels_dbscan = dbscan.fit_predict(X_scaled)
df["DBSCAN_Cluster"] = labels_dbscan

print("Unique Labels:", np.unique(labels_dbscan))
print("Clusters:", len(set(labels_dbscan)) - (1 if -1 in labels_dbscan else 0))
print("Noise Points:", list(labels_dbscan).count(-1))

Unique Labels: [-1  0  1]
Clusters: 2
Noise Points: 26


In [45]:
fig = go.Figure()

for label in np.unique(labels_dbscan):
    if label == -1:
        fig.add_trace(go.Scatter(
            x=X[labels_dbscan == label, 0],
            y=X[labels_dbscan == label, 1],
            mode='markers',
            name='Noise',
            marker=dict(color='black', size=6, symbol='x'),
            hovertemplate='<b>Noise</b><br><b>Annual Income:</b> %{x}k$<br><b>Spending Score:</b> %{y}<extra></extra>'
        ))
    else:
        fig.add_trace(go.Scatter(
            x=X[labels_dbscan == label, 0],
            y=X[labels_dbscan == label, 1],
            mode='markers',
            name=f'Cluster {label}',
            marker=dict(size=8),
            hovertemplate=f'<b>Cluster {label}</b><br><b>Annual Income:</b> %{{x}}k$<br><b>Spending Score:</b> %{{y}}<extra></extra>'
        ))

fig.update_layout(
    title='DBSCAN Clustering (eps=0.2)',
    xaxis_title='Annual Income (k$)',
    yaxis_title='Spending Score (1-100)',
    template='plotly_white',
    hovermode='closest',
    width=650,
    height=400
)

fig.show()

In [46]:
mask = labels_dbscan != -1
if len(set(labels_dbscan[mask])) > 1:
    score = silhouette_score(X_scaled[mask], labels_dbscan[mask])
    print(f"Silhouette Score: {score:.4f}")
else:
    print("Not enough clusters for Silhouette Score")

Silhouette Score: 0.1654


## 9. Model Comparison
### 9.1 Silhouette Score Comparison

We compare all clustering algorithms based on Silhouette Score
and number of clusters to determine the best performing model
for mall customer segmentation.



In [47]:
from time import time

results = []

# K-Means
t = time()
km = KMeans(n_clusters=5, random_state=42, n_init=10, init='random')
l = km.fit_predict(X_scaled)
results.append({
    "Algorithm": "K-Means",
    "N Clusters": 5,
    "Silhouette": round(silhouette_score(X_scaled, l), 4),
    "Runtime (s)": round(time() - t, 3)
})


# Hierarchical
t = time()
h = AgglomerativeClustering(n_clusters=5, linkage='ward')
l = h.fit_predict(X_scaled)
results.append({
    "Algorithm": "Hierarchical",
    "N Clusters": 5,
    "Silhouette": round(silhouette_score(X_scaled, l), 4),
    "Runtime (s)": round(time() - t, 3)
})

# DBSCAN
t = time()
db = DBSCAN(eps=0.2, min_samples=5)
l = db.fit_predict(X_scaled)
mask = l != -1
n_clusters = len(set(l)) - (1 if -1 in l else 0)
sil = round(silhouette_score(X_scaled[mask], l[mask]), 4) if n_clusters > 1 else "N/A"
results.append({
    "Algorithm": "DBSCAN",
    "N Clusters": n_clusters,
    "Silhouette": sil,
    "Runtime (s)": round(time() - t, 3)
})



df_results = pd.DataFrame(results)

In [48]:
print("Clustering Algorithms Performance Comparison:")
df_results

Clustering Algorithms Performance Comparison:


,Algorithm,N Clusters,Silhouette,Runtime (s)
0,K-Means,5,0.3689,0.138
1,Hierarchical,5,0.3112,0.407
2,DBSCAN,2,0.1654,0.089


In [49]:
df_plot = df_results[df_results["Silhouette"] != "N/A"].copy()
df_plot["Silhouette"] = df_plot["Silhouette"].astype(float)
best = df_plot["Silhouette"].idxmax()

colors = ['#E74C3C' if i == best else '#3498DB' for i in df_plot.index]

fig = go.Figure()

fig.add_trace(go.Bar(
    x=df_plot["Algorithm"],
    y=df_plot["Silhouette"],
    marker=dict(
        color=colors,
        line=dict(color='black', width=1)
    ),
    hovertemplate='%{x}Silhouette Score: %{y:.4f}'
))

fig.update_layout(
    title='Silhouette Score Comparison — Best Algorithm in Red',
    xaxis_title='Algorithm',
    yaxis_title='Silhouette Score',
    template='plotly_white',
    width=700,
    height=400
)

fig.show()

### 9.2 Additional Evaluation Metrics and Scalability

The earlier comparison used silhouette score and runtime. To strengthen the evaluation, we add Calinski-Harabasz and Davies-Bouldin scores to assess cluster separation and compactness. We also benchmark runtime across different sample sizes to discuss scalability.

In [50]:
from time import perf_counter


def compute_metrics(labels, X, algorithm_name):
    mask = labels != -1 if algorithm_name == "DBSCAN" else np.ones(len(labels), dtype=bool)
    X_use = X[mask]
    labels_use = labels[mask]

    if len(np.unique(labels_use)) < 2:
        return {
            "Algorithm": algorithm_name,
            "N Clusters": len(np.unique(labels_use)),
            "Silhouette": np.nan,
            "Calinski-Harabasz": np.nan,
            "Davies-Bouldin": np.nan,
        }

    return {
        "Algorithm": algorithm_name,
        "N Clusters": len(np.unique(labels_use)),
        "Silhouette": round(silhouette_score(X_use, labels_use), 4),
        "Calinski-Harabasz": round(calinski_harabasz_score(X_use, labels_use), 2),
        "Davies-Bouldin": round(davies_bouldin_score(X_use, labels_use), 2),
    }


evaluation_metrics = pd.DataFrame([
    compute_metrics(labels_kmeans, X_scaled, "K-Means"),
    compute_metrics(labels_hier, X_scaled, "Hierarchical"),
    compute_metrics(labels_dbscan, X_scaled, "DBSCAN"),
])

evaluation_metrics

,Algorithm,N Clusters,Silhouette,Calinski-Harabasz,Davies-Bouldin
0,K-Means,5,0.3636,1867.86,0.89
1,Hierarchical,5,0.3112,1597.71,0.95
2,DBSCAN,2,0.1654,8.81,0.69


In [51]:
for metric in ["Calinski-Harabasz", "Davies-Bouldin"]:
    plot_df = evaluation_metrics[["Algorithm", metric]].copy().dropna()

    if plot_df.empty:
        continue

    fig = px.bar(
        plot_df,
        x="Algorithm",
        y=metric,
        color="Algorithm",
        title=f"{metric} Comparison",
        labels={metric: metric},
        template="plotly_white",
        text=plot_df[metric].round(2)
    )
    fig.update_traces(textposition="outside")
    fig.update_layout(
        showlegend=False,
        width=640,
        height=400,
        margin=dict(t=50, l=20, r=20, b=20),
        yaxis=dict(range=[0, max(plot_df[metric].max() * 1.15, 1.0)])
    )
    fig.show()

In [52]:
sample_sizes = [200, 500, 1000]
scalability_results = []

for n in sample_sizes:
    X_sub = X_scaled[:n]
    for algorithm_name, model in [
        ("K-Means", KMeans(n_clusters=5, random_state=42, n_init=10)),
        ("Hierarchical", AgglomerativeClustering(n_clusters=5)),
        ("DBSCAN", DBSCAN(eps=0.2, min_samples=5)),
    ]:
        start = perf_counter()
        model.fit_predict(X_sub)
        runtime = perf_counter() - start
        scalability_results.append({
            "Algorithm": algorithm_name,
            "Sample Size": n,
            "Runtime (s)": round(runtime, 4),
        })

scalability_df = pd.DataFrame(scalability_results)
scalability_df

,Algorithm,Sample Size,Runtime (s)
0,K-Means,200,0.0376
1,Hierarchical,200,0.0021
2,DBSCAN,200,0.0043
3,K-Means,500,0.0364
4,Hierarchical,500,0.0075
5,DBSCAN,500,0.0077
6,K-Means,1000,0.0385
7,Hierarchical,1000,0.0263
8,DBSCAN,1000,0.0209


## 10. Customer Segmentation & Business Insights

Based on K-Means clustering results, we profile each customer segment using their average Age, Annual Income, and Spending Score.

In [53]:
cluster_profile = df.groupby("KMeans_Cluster")[["Age", "AnnualIncome", "SpendingScore"]].mean().round(2)
cluster_profile

,Age,AnnualIncome,SpendingScore
KMeans_Cluster,,,
0,28.02,29.80,66.77
1,38.68,61.46,42.24
2,46.62,84.55,26.83
3,31.13,37.92,51.37
4,66.17,107.92,4.41


In [54]:
cluster_profile_norm = cluster_profile.copy()
for col in cluster_profile_norm.columns:
    cluster_profile_norm[col] = (
        (cluster_profile_norm[col] - cluster_profile_norm[col].min()) /
        (cluster_profile_norm[col].max() - cluster_profile_norm[col].min())
    )

fig = px.imshow(
    cluster_profile_norm,
    text_auto=True,
    color_continuous_scale='RdYlGn',
    title='Cluster Characteristics Heatmap (Normalized)',
    labels=dict(x='Features', y='Cluster', color='Normalized Value')
)

fig.update_layout(
    xaxis_title='Features',
    yaxis_title='Cluster',
    width=640,
    height=500,
    template='plotly_white'
)
fig.show()



## 11. Conclusion 

### 11.1 Summary of Findings

This project performed a comprehensive customer segmentation
analysis on 982 mall customers using 3 different clustering
algorithms. Key findings are summarized below.

**Dataset:**
- 982 customers after cleaning
- Features used: AnnualIncome, SpendingScore
- Optimal clusters: K = 5


### 11.2 Best Algorithm

Based on Silhouette Score and cluster interpretability,
**K-Means** performed best for this dataset because:
- Clear, well-separated clusters
- Fast and scalable
- Highly interpretable results
- Stable across multiple runs

### 11.3 Final Comparison Table

| Algorithm | Strengths | Weaknesses | Suitability for This Dataset |
|-----------|-----------|------------|------------------------------|
| K-Means | Fast, simple, scalable, and easy to interpret | Assumes spherical clusters and requires a predefined number of clusters | **Chosen model** because the resulting clusters were clear and well separated |
| Hierarchical Clustering | Does not require specifying K in advance and provides a dendrogram for structure | More computationally expensive and less scalable for large datasets | Useful for understanding relationships between groups, but less practical here than K-Means |
| DBSCAN | Finds arbitrary-shaped clusters and can detect outliers | Sensitive to parameter tuning and struggles with varying-density clusters | Less suitable here because the mall customer data is relatively uniform and dense |

### 11.4 Key Insights and Scalability Considerations

**How the Customer Segments Were Identified:**

The customer segments were derived by applying K-Means clustering to the standardized features of Annual Income and Spending Score. After selecting K = 5 using the elbow method and silhouette analysis, the algorithm grouped customers into five clusters based on similarity in their spending and income behavior. Each cluster was then interpreted by examining the average characteristics of its members, such as income level, spending pattern, and age profile.

**Customer Segments Identified:**

- **Cluster 0** — Young low-income high spenders → Best ROI
- **Cluster 1** — Middle-aged medium income moderate spenders → Stable revenue
- **Cluster 2** — Mature high income low spenders → Needs engagement
- **Cluster 3** — Young low-medium income moderate spenders → Growth potential
- **Cluster 4** — Senior highest income non-spenders → Re-engagement needed

**Algorithm Comparison and Scalability:**
- **K-Means** — Best overall choice for this dataset because it produced clear clusters, was fast to train, and scaled well with larger sample sizes.
- **Hierarchical Clustering** — Useful for understanding cluster relationships, but it becomes less practical for large datasets because its computational cost grows quickly.
- **DBSCAN** — Strong for detecting outliers and non-spherical clusters, but it was less suitable here because the parameter selection was sensitive and its performance was less stable for this relatively dense dataset.

Overall, the results suggest that K-Means is the most practical and business-friendly approach for mall customer segmentation, especially when scalability and interpretability are important.